_posts/2020-10-07-torch-modules/torch_modules.Rmd

```diff
@@ -0,0 +1,429 @@
+say --, as we've been doing so far, we can create a linear module. The
+following snippet instantiates a linear layer that expects three-feature
+inputs and returns a single output per observation:
+
+```{r}
+library(torch)
+l <- nn_linear(3, 1)
+```
+
+The module has two parameters, "weight" and "bias". Both have already
+been initialized:
+
+```{r}
+l$parameters
+```
+
+    $weight
+    torch_tensor 
+    -0.0385  0.1412 -0.5436
+    [ CPUFloatType{1,3} ]
+
+    $bias
+    torch_tensor 
+    -0.1950
+    [ CPUFloatType{1} ]
+
+Modules are callable; calling a module executes its `forward()` method,
+which for a linear layer just matrix-multiplies input and weights and
+adds the bias.
+
+Let's try this:
+
+```{r}
+data  <- torch_randn(10, 3)
+out <- l(data)
+```
+
+Unsurprisingly, `out` now holds some data:
+
+```{r}
+out$data()
+```
+
+    torch_tensor 
+     0.2711
+    -1.8151
+    -0.0073
+     0.1876
+    -0.0930
+     0.7498
+    -0.2332
+    -0.0428
+     0.3849
+    -0.2618
+    [ CPUFloatType{10,1} ]
+
+In addition though, it knows what will need to be done, should ever it
+be asked to calculate gradients:
+
+```{r}
+out$grad_fn
+```
+
+    AddmmBackward
+
+(You may remember from the preceding post that when we work with tensors
+directly, we need to create them passing `requires_grad = TRUE` to
+achieve this. With modules, `torch` correctly assumes that we'll want to
+perform backpropagation at some point.)
+
+By now though, we haven't called `backward()` yet. Thus, no gradients
+have yet been calculated:
+
+```{r}
+l$weight$grad
+l$bias$grad
+```
+
+    torch_tensor 
+    [ Tensor (undefined) ]
+    torch_tensor 
+    [ Tensor (undefined) ]
+
+Let's change this:
+
+```{r, error = TRUE}
+out$backward()
+```
+
+    Error in (function (self, gradient, keep_graph, create_graph)  : 
+      grad can be implicitly created only for scalar outputs (_make_grads at ../torch/csrc/autograd/autograd.cpp:47)
+
+Why the error? *Autograd* expects the output tensor to be a scalar,
+while in our example, we have a tensor of size `(10, 1)`. This error
+won't often occur in practice, where we work with *batches* of inputs
+(sometimes, just a single batch). But still, it's interesting to see how
+to resolve this.
+
+To make the example work, we introduce a -- virtual -- final aggregation
+step -- taking the mean, say. Let's call it `avg`. If such a mean were
+taken, its gradient with respect to `l$weight` would be obtained via the
+chain rule:
+
+$$
+\begin{equation*} 
+ \frac{\partial \ avg}{\partial w} = \frac{\partial \ avg}{\partial \ out}  \ \frac{\partial \ out}{\partial w}
+\end{equation*}
+$$
+
+Of the quantities on the right side, we're interested in the second. We
+need to provide the first one, the way it would look *if really we were
+taking the mean*:
+
+```{r}
+d_avg_d_out <- torch_tensor(10)$`repeat`(10)$unsqueeze(1)$t()
+out$backward(gradient = d_avg_d_out)
+```
+
+Now, `l$weight$grad` and `l$bias$grad` *do* contain the respective
```

Comment:
```suggestion
Now, `l$weight$grad` and `l$bias$grad` *do* contain their respective
```

Reply:
yeah, right... actually i'm probably being overly clear here as often ... `do contain gradients` should suffice...